### ISSUE TITLE
Unexpected behavior with __CPROVER_assume and circular data structures

### ISSUE THREAD

**Issue by markrtuttle** [diffblue/cbmc#5004]
CBMC version: 5.12 (cbmc-5.12-d8598f8-767-g06ef36d45)
Operating system: macOS Sierra 10.12.6
Exact command line resulting in the issue: goto-cc demo.c; cbmc --trace a.out
What behaviour did you expect: CBMC would succeed
What happened instead: CBMC fails with a trace of an assertion failure.

The attached file [demo.c.zip](https://github.com/diffblue/cbmc/files/3490078/demo.c.zip) demonstrates what looks like a bug involving the interaction of __CPROVER_assume and function invocation and circular data structures.

If I run ```goto-cc demo.c; cbmc --trace a.out``` I get a trace for an assertion failure at the end of main.  Embedded in the code is a change that makes the problem go away.  If I run ```goto-cc -DCHOICE demo.c; cbmc a.out``` the assertion succeeds.

The code involves a predicate describing circularity

```
int predicate(struct structB *x) {
  return x->structB_member1 == x;
}
```

and a function

```
void function (struct structB  *structb1){
  struct structB  *structb2 = (struct structB *)(structb1->structB_member1);
  structb2->structB_member2 = malloc(10);
}
```
and a main

```
int main() {
  struct structB *STRUCTB = malloc(sizeof(struct structB));
  __CPROVER_assume(predicate(STRUCTB));
  function(STRUCTB->structB_member1);
  assert(STRUCTB->structB_member2);
}
```

Since STRUCTB and STRUCTB->structB_member1 and structb1 and structb2 all point to  STRUCTB, I expect the assignment to structb2->structB_member2 to update STRUCTB->structB_member2.  But the assertion fails, demonstrating that this doesn't happen.

If I replace the function invocation ```__CPROVER_assume(predicate(STRUCTB));``` with the straight assumption ```__CPROVER_assume(STRUCTB->structB_member1 == STRUCTB);``` then the proof succeeds.  This is what building with -DCHOICE is doing.

If I add an ```assert(0);``` to the end of main to ensure trace generation, the only difference in the traces for the ```assert(0)``` failure is the assignment in line 16 in the function.  With -DCHOICE the line is what you expect.

```
dynamic_object1.structB_member2=(char *)dynamic_object2
```

Without -DCHOICE (in the demo itself) the line is

```
structb2$object.structB_member2=(char *)dynamic_object2
```

**Comment by tautschnig** [comment]
@smowton Would you be able to take a look at this? The problem is (also) reproducible by using `--no-propagation` and really seems to boil down to symex not having any candidate values in the value set for `structb1` in `struct structB  *structb2 = (struct structB *)(structb1->structB_member1);` - which is not too surprising given that there is no assignment to the underlying object.

**Comment by smowton** [comment]
Acknowledged, I do intend to look at this though may not find time before my holiday 9th-20th September.

**Comment by smowton** [comment]
OK, confirmed, the cause is that the value of `STRUCTB->structB_member1` is never explicitly assigned, only implied via an assumption. That means that while solver figures out what it must be, the value-set (alias analysis) used in lowering pointer operations into concrete object references doesn't know where it might point, and offers `structb2$object` as a stand-in for the unknown pointer value.

Using `__CPROVER_assume(STRUCTB->structB_member1 == STRUCTB);` stumbles by happenstance into a particular special case which populates the value-set and constant propagator when faced with an assumption of the form `symbol == constant` (in this case, `dynamic_object1..structB_member1 == &dynamic_object1`, after dereferencing has been applied). By contrast `ret = dynamic_object1..structB_member1 == &dynamic_object1; ASSUME ret` isn't good enough to get noticed.

I think the only way to solve this sort of thing in the general case would be to produce a very large formula when our alias analysis fails which considers every value the nondet pointer might have -- i.e.  every possible offset into every possible object. If you want to split the difference between fidelity and cost you could go for every field-aligned offset, or every object root, or only objects of the appropriate type. Since dynamic objects and locals are produced throughout symex's run you would have to emit some "unknown pointer" symbol to fill in post-symex with a switch-case over the whole universe.

In summary: known design limitation, so this is a feature request (support pointer aliases implied by indirect assumption) rather than a bug.

**Comment by feliperodri** [comment]
@smowton @tautschnig @markrtuttle any updates on this issue? Should we close it?

**Comment by hannes-steffenhagen-diffblue** [comment]
`__CPROVER_assume` can only be used to "prune" possible nondeterministic states. It can't make anything true as is the expectation in the bug report here. This is unlikely to change so closing for now. Feel free to reopen if you believe this is in error.
